### DISCUSSION TITLE
Can I tell a grammar about exception handling?

### DISCUSSION THREAD

**Discussion by YPTopMan** [dotnet/csharplang#2581] [Language Ideas]
![image](https://user-images.githubusercontent.com/11879898/58850829-baf66c80-86c2-11e9-917e-74d9c23d66c9.png)

There may be thousands of such judgments in the project, and then an exception is thrown. Can it be extracted into a single code to do it?

Ta like this, Knowing that it is a judgment parameter, but it takes up the code line, you need to scroll the mouse, will it affect the view of the main code?
![image](https://user-images.githubusercontent.com/11879898/58850939-466ffd80-86c3-11e9-9400-a7ce90347e6d.png)

![image](https://user-images.githubusercontent.com/11879898/58851038-bed6be80-86c3-11e9-80b0-4c63adce139e.png)

If the line can do it, you can save a lot of if (){}
![image](https://user-images.githubusercontent.com/11879898/58851074-e037aa80-86c3-11e9-980c-4e866dbda16d.png)

**Comment by CyrusNajmabadi** [comment]
`if (name == null) throw new ArgumentNullException(nameof(name))`

However, we're also adding the following syntax in the future to do the same:

```c#
void Foo(string name!)  // <-- note the !
{
}
```

**Comment by yaakov-h** [comment]
Also valid:

```csharp
_ = name ?? throw new ArgumentNullException(nameof(name));
```

**Comment by YPTopMan** [comment]
> Also valid:
> 
> ```cs
> _ = name ?? throw new ArgumentNullException(nameof(name));
> ```

![image](https://user-images.githubusercontent.com/11879898/58855847-aa022700-86d3-11e9-866f-9c5a6d1a194c.png)

The code behind it is grayed out, it feels a bit strange, I hope to have support for this exception handling.

**Comment by theunrepentantgeek** [comment]
The emerging model that I've seen in a bunch of code is to move the checks inline

``` csharp
return ModelBindingsHelper.TryUpdateModelAsync(
    model ?? throw new ArgumentNullException(nameof(model)),
    prefix,
    ControllerContext,
    MetadataProvider,
    ModelBinderFactory,
    valueProvider ?? throw new ArgumentNullException(nameof(valueProvider)),
    ObjectValidator,
    includeExpressions ?? throw new ArgumentNullException(nameof(includeExpressions)));
```

This saves just as much code - and can be written **now**, no requirement for language or compiler changes.

**Comment by YPTopMan** [comment]
> `if (name == null) throw new ArgumentNullException(nameof(name))`
> 
> However, we're also adding the following syntax in the future to do the same:
> 
> ```cs
> void Foo(string name!)  // <-- note the !
> {
> }
> ```


This is less {}, although it is on one line, but it is not suitable for large-scale implementation.




void Foo(string name!)  // <-- note the !
{
}
Custom processing methods can also be done, but you need to extend a class library yourself, so that each project may have a similar duplicate library.

**Comment by CyrusNajmabadi** [comment]
> This is less {}, although it is on one line, but it is not suitable for large-scale implementation.

Why is it not suitable?

**Comment by CyrusNajmabadi** [comment]
> Custom processing methods can also be done, but you need to extend a class library yourself, so that each project may have a similar duplicate library.

I don't know what this means.

**Comment by YPTopMan** [comment]
> > This is less {}, although it is on one line, but it is not suitable for large-scale implementation.
> 
> Why is it not suitable?
Many people like to use {} to express, do not like to remove {}, and remove {}, will make the style of the code into two styles, if there is direct support, it will not consider whether to {}

**Comment by YPTopMan** [comment]
> The emerging model that I've seen in a bunch of code is to move the checks inline
> 
> ```cs
> return ModelBindingsHelper.TryUpdateModelAsync(
>     model ?? throw new ArgumentNullException(nameof(model)),
>     prefix,
>     ControllerContext,
>     MetadataProvider,
>     ModelBinderFactory,
>     valueProvider ?? throw new ArgumentNullException(nameof(valueProvider)),
>     ObjectValidator,
>     includeExpressions ?? throw new ArgumentNullException(nameof(includeExpressions)));
> ```
> 
> This saves just as much code - and can be written **now**, no requirement for language or compiler changes.


Yes, this is fine, but it won't work in the body of the method, because a variable must be used to receive the normal situation.

**Comment by theunrepentantgeek** [comment]
@YPTopMan, I think you've missed something.

```
void Foo(string name!)  // <-- note the !
{
}
```

This syntax is about enforcing not-null in the _signature of the method_ itself.

It's not a separate statement within existing code.

**Comment by CyrusNajmabadi** [comment]
> Many people like to use {} to express, do not like to remove {}, and remove {}, will make the style of the code into two styles, if there is direct support, it will not consider whether to {}

Then write it as:

```c#
if (name == null) { throw new ArgumentNullException(nameof(name)); }
```

**Comment by theunrepentantgeek** [comment]
> a variable must be used to receive the normal situation.

Incorrect. `expr ?? throw ...` is just an expression - and can be used anywhere you put an expression.

**Comment by CyrusNajmabadi** [comment]
You complained about it being 4 lines.  But it's only four lines because you spread it over four lines.  If you want it on one line, you don't need a new construct, you can just literally place the code on one line.  It's not necessary to add new grammar constructs to make things brief just because you don't want to use the existing constructs that can already be brief :)

**Comment by CyrusNajmabadi** [comment]
> Yes, this is fine, but it won't work in the body of the method, because a variable must be used to receive the normal situation.

This is not the case.  You can use that form anywhere an expression would be allowed.

**Comment by CyrusNajmabadi** [comment]
@YPTopMan you have mixed up the quoting and code-blocks in several of your posts.  You can use ` ```c# ` to start a code-block and then ` ``` ` to end one.

You can also edit your posts after the fact to fixup these issues.

**Comment by YPTopMan** [comment]
> > Custom processing methods can also be done, but you need to extend a class library yourself, so that each project may have a similar duplicate library.
> 
> I don't know what this means.



> > a variable must be used to receive the normal situation.
> 
> Incorrect. `expr ?? throw ...` is just an expression - and can be used anywhere you put an expression.

![image](https://user-images.githubusercontent.com/11879898/58857288-63aec700-86d7-11e9-9335-9ef2a76e8229.png)

**Comment by YPTopMan** [comment]
> You complained about it being 4 lines. But it's only four lines because you spread it over four lines. If you want it on one line, you don't need a new construct, you can just literally place the code on one line. It's not necessary to add new grammar constructs to make things brief just because you don't want to use the existing constructs that can already be brief :)

![image](https://user-images.githubusercontent.com/11879898/58857393-bbe5c900-86d7-11e9-936c-abd85dcbfee7.png)

This is the source code of AspNetCore. Once you enter the method, it is the judgment parameter. You need to drag the mouse to see the real processing logic.

**Comment by CyrusNajmabadi** [comment]
>This is the source code of AspNetCore. Once you enter the method, it is the judgment parameter. You need to drag the mouse to see the real processing logic.

And that's their choice.  They could write it in a single line as:

```c#
if (name == null) { throw new ArgumentNullException(nameof(name)); }
```

Or, in a future version of C# as 

```c#
void Whatever(Model model!, string modelType!, Provider valueProvider!, Filter propertyFilter!)
{
}
```

**Comment by CyrusNajmabadi** [comment]
> This is the source code of AspNetCore.

Then ask them to change their code.  Even if C# adds a new language feature, it won't make this any better unless they change their code.

unless you are asking for some sort of IDE feature to collapse things here?  In which case, you should be clear what you'd like from the IDE, and you should move this issue to dotnet/roslyn, not dotnet/csharplang.

**Comment by CyrusNajmabadi** [comment]
![image](https://user-images.githubusercontent.com/4564579/58857972-57852400-865b-11e9-95ea-7ef9241a61ec.png)


As stated, you will be able to write that in the future as:

```c#
public string Get(EntityResult model!) // <-- again, note the !
{
    return model.B;
}
```

And that's all you'll need to write.  You asked for a grammar change to make that simpler to write.  And it doesn't get much simpelr than `!` :)

**Comment by YPTopMan** [comment]
> > This is the source code of AspNetCore.
> 
> Then ask them to change their code. Even if C# adds a new language feature, it won't make this any better unless they change their code.
> 
> unless you are asking for some sort of IDE feature to collapse things here? In which case, you should be clear what you'd like from the IDE, and you should move this issue to dotnet/roslyn, not dotnet/csharplang.

Yes, no one will change it. But the code is more streamlined. The person reading the source code may increase. Everyone can modify a little bit, or make a little suggestion. Isn't it better? I also got the source code yesterday. Of course, this is what I personally feel can be improved.

**Comment by YPTopMan** [comment]
> ![image](https://user-images.githubusercontent.com/4564579/58857972-57852400-865b-11e9-95ea-7ef9241a61ec.png)
> 
> As stated, you will be able to write that in the future as:
> 
> ```cs
> public string Get(EntityResult model!) // <-- again, note the !
> {
>     return model.B;
> }
> ```
> 
> And that's all you'll need to write. You asked for a grammar change to make that simpler to write. And it doesn't get much simpelr than `!` :)

![image](https://user-images.githubusercontent.com/11879898/58858489-357eb680-86da-11e9-82de-02583fcb7f34.png)

Compilation failed

**Comment by CyrusNajmabadi** [comment]
> Everyone can modify a little bit, or make a little suggestion. Isn't it better?

It's unclear what sort of suggestion you're asking for.  

1. Are you asking for Asp.net to update their code to make it more streamlined?  If so, that feedback should go to them (possibly with a PR you create).  And in this case `dotnet/csharplang` is not the right repro for this issue.
2. Or, are you asking for the language to have options to make it easier to write this without taking up so much space?  If so, there have been three ways listed in the language to make this better.  There's no need for yet *another* way when these 3 other ways already address the problem.
3. Or, are you asking for the IDE to provide a way to 'streamline' things when you're looking at this code?  In that case, this isn't appropriate for `dotnet/csharplang` and should go to `dotnet/roslyn`.

Which of the above are you asking for?  Your posts don't make it clear at all.

**Comment by CyrusNajmabadi** [comment]
> Compilation failed

> "As stated, you will be able to write that in the **future** as:"

i.e. the language is already adding this as a **future** feature.  So there's no need for **another** feature to do the exact same thing since the `!` feature is already being worked on.

https://github.com/dotnet/roslyn/tree/features/param-nullchecking is the branch where this is being worked on.  It will be in a future version of C#.

**Comment by YPTopMan** [comment]
> > Everyone can modify a little bit, or make a little suggestion. Isn't it better?
> 
> It's unclear what sort of suggestion you're asking for.
> 
> 1. Are you asking for Asp.net to update their code to make it more streamlined?  If so, that feedback should go to them (possibly with a PR you create).  And in this case `dotnet/csharplang` is not the right repro for this issue.
> 2. Or, are you asking for the language to have options to make it easier to write this without taking up so much space?  If so, there have been three ways listed in the language to make this better.  There's no need for yet _another_ way when these 3 other ways already address the problem.
> 3. Or, are you asking for the IDE to provide a way to 'streamline' things when you're looking at this code?  In that case, this isn't appropriate for `dotnet/csharplang` and should go to `dotnet/roslyn`.
> 
> Which of the above are you asking for? Your posts don't make it clear at all.

I am hoping that the exception is supported by the following format?
![image](https://user-images.githubusercontent.com/11879898/58859329-12ed9d00-86dc-11e9-89b3-dd8a2c0ea1a6.png)

**Comment by YPTopMan** [comment]
> > Compilation failed
> 
> > "As stated, you will be able to write that in the **future** as:"
> 
> i.e. the language is already adding this as a **future** feature. So there's no need for **another** feature to do the exact same thing since the `!` feature is already being worked on.
> 
> https://github.com/dotnet/roslyn/tree/features/param-nullchecking is the branch where this is being worked on. It will be in a future version of C#.

![image](https://user-images.githubusercontent.com/11879898/58859375-3284c580-86dc-11e9-8a59-892261f10a1e.png)


I came from this link, I thought it was necessary to design support? Sorry

**Comment by CyrusNajmabadi** [comment]
> I am hoping that the exception is supported by the following format?

It will not be.  The formats that will be supported are:

```c#
void Foo(Model model!)
{
}

// or

some_expression = model ?? throw Whatever(...);

// or

DoSomething(model ?? throw Whatever(...));

// or

if (model is null) { throw Whatever(...); }
```

(And the 2nd and 3rd are really the same thing).  The language isn't going to add a 5th form when there are already four ways to do waht you're asking for.

**Comment by Dersei** [comment]
About this:
```csharp
_ = name ?? throw new ArgumentNullException(nameof(name));
```
The part after `??` is grayed out in your screenshot, because you have already checked this condition before and `name` is never a null. If you remove `if` it shouldn't be grayed out any longer.

**Comment by CyrusNajmabadi** [comment]
>I came from this link, I thought it was necessary to design support? Sorry

Yes, if you want to discuss language change proposals, you're in the right place.  However, it's highly unlikely the language will take the naked `name ?? throw` form, since there are 4 other ways you can do this already today.

**Comment by The-Futurist** [comment]
@YPTopMan - Consider leveraging some static guard support. I've done this recently more and more and its perhaps the best way so far that we've approached this.

There are many [examples](https://github.com/ardalis/GuardClauses) around and you can readily code your own.

You can also design these so that the caller passes a lambda in, you end up with a visually very uniform set of opening statements in your methods. This eases code review, readability and maintenance. 

You basically end up with one line per argument.

Finally the "[using static](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static)" feature (which I love) means you can even remove the static container class's name.

Anyway, here's the idea for all this:

```cs
    public static class GuardMethods
    {
        public static void ThrowIfNull(string ArgName, object Arg)
        {
            ThrowIfArgNameIsNull(ArgName);
            if (Arg == null) throw new ArgumentNullException(ArgName);
        }

        public static void ThrowIfEmpty<T>(string ArgName, IEnumerable<T> Sequence)
        {
            ThrowIfArgNameIsNull(ArgName);
            ThrowIfNull(nameof(Sequence), Sequence);

            if (!Sequence.Any()) throw new ArgumentException("The sequence must not be empty.", ArgName);
        }

        private static void ThrowIfArgNameIsNull(string ArgName)
        {
            if (ArgName == null) throw new InvalidOperationException("Improper use of guard library. Argument name must never be null.");
        }
    }

```

**Comment by YPTopMan** [comment]
Thank you all
